README.md
This README is just a quick start document. You can find more detailed documentation of Spunky Bot at [https://spunkybot.de](https://spunkybot.de).

# What is Spunky Bot?

**Spunky Bot** is a lightweight game server administration bot and RCON tool.
Its purpose is to administer, manage and maintain an [Urban Terror](http://www.urbanterror.info) 4.1 / 4.2 / 4.3 server and to provide real time statistics data for players.
Spunky Bot is a cross-platform package and offers in-game commands without authentication and automated administration even when no admin is online.
The code of Spunky Bot is inspired by the eb2k9 bot by Shawn Haggard, which was released under the Beerware License.

[![Build Status](https://travis-ci.org/SpunkyBot/spunkybot.png)](https://travis-ci.org/SpunkyBot/spunkybot)
[![Actions Status](https://github.com/SpunkyBot/spunkybot/workflows/CI/badge.svg)](https://github.com/SpunkyBot/spunkybot/actions)
[![License](https://img.shields.io/github/license/SpunkyBot/spunkybot)](https://github.com/SpunkyBot/spunkybot/blob/develop/LICENSE)
[![Release](https://img.shields.io/github/v/release/SpunkyBot/spunkybot.svg?color=orange)](https://github.com/SpunkyBot/spunkybot/releases)
[![PyPI version](https://img.shields.io/pypi/v/spunkybot.svg)](https://pypi.python.org/pypi/spunkybot)
[![Python version](https://img.shields.io/pypi/pyversions/spunkybot?color=yellow)](https://pypi.org/project/spunkybot)
[![GitHub Downloads](https://img.shields.io/github/downloads/SpunkyBot/spunkybot/total?color=yellowgreen)](https://github.com/SpunkyBot/spunkybot/releases)

If you want to know more, this is a list of selected starting points:

* Introduction to [Spunky Bot](https://spunkybot.de)
* The full list of [commands](https://spunkybot.de/docs/commands)
* There is much more inside the [official documentation](https://docs.spunkybot.de)

## Features

* Lightweight and fast
* Real time game statistics
* Different user groups and levels
* Supports all RCON commands
* Supports temporary and permanent bans of players
* Supports rotation messages
* Stores all player related information in a SQLite database
* Runs 'out of the box', no other software requirements

## Environment

* Urban Terror 4.1.1 / 4.2.023 / 4.3.4
* Python 2.6 / 2.7
* SQLite 3 database
* Cross-platform (tested on Debian 6 - 10, Ubuntu 10.04 - 20.04, CentOS 6 / 7, macOS 10.13, Windows 7 / 10)
* Supporting 32-bit and 64-bit operating systems

## Quickstart

It's easy to get started with Spunky Bot:

* Download the [latest release](https://spunkybot.de/download)
* Read our [quickstart guide](https://spunkybot.de/docs/start) to get up and running

### Configuration

* Modify the Urban Terror server config file as follows:

```
seta g_logsync "1"
seta g_loghits "1"
seta g_friendlyfire "2"
```

* Restart your Urban Terror server
* Modify the Spunky Bot configuration file `/conf/settings.conf` and set game server port and RCON password
* In-game displayed rules/advertisements are contained in the file `/conf/rules.conf`
* If you do not want to display the rotation messages, set the value `show_rules=0` in the config file `/conf/settings.conf`
* Run the application manually: `python spunky.py`
* Or use the provided systemd or sysVinit script to run Spunky Bot as daemon

**_First start instruction:_**

* Connect to your game server and type `!iamgod` in the global chat to get the admin level "Head Admin". This command is only once available.

## Documentation

You can find all details in the latest [documentation](https://docs.spunkybot.de).

### Bot Commands

The description of all available [commands](https://github.com/SpunkyBot/spunkybot/blob/develop/doc/Commands.md) as well as the admin levels and rights is located under the subfolder `/doc`.

## Resources

* [Documentation](https://docs.spunkybot.de)
* [Bug Tracker](https://github.com/SpunkyBot/spunkybot/issues)
* [Mailing List](https://groups.google.com/group/spunkybot)
* [Source Code](https://github.com/SpunkyBot/spunkybot)
* [Homepage](https://spunkybot.de)

## Changelog

You can keep up-to-date with the changes that we have made via our [releases page](https://github.com/Spunkybot/spunkybot/releases).

## Versioning

Spunky Bot is currently maintained under the [Semantic Versioning](http://semver.org) guidelines. Releases will be numbered with the following format: `<major>.<minor>.<patch>`

## Additional Information

For additional information, visit the Spunky Bot website at [https://www.spunkybot.de](https://www.spunkybot.de).

If you have any questions about Spunky Bot or need help, please use the [mailing list](https://groups.google.com/group/spunkybot).

* Joining the mailing list:
  * Send an email to <mailto:spunkybot+subscribe@googlegroups.com>
  * or subscribe via web: [https://groups.google.com/forum/#!forum/spunkybot/join](https://groups.google.com/forum/#!forum/spunkybot/join)
* Leaving the mailing list:
  * Send an email to <mailto:spunkybot+unsubscribe@googlegroups.com>
  * or unsubscribe via web: [https://groups.google.com/forum/#!forum/spunkybot/unsubscribe](https://groups.google.com/forum/#!forum/spunkybot/unsubscribe)
* If you joined the group, you can post to this group:
  * Send an email to spunkybot@googlegroups.com

If you have bug reports or feature suggestions, please use the [issue tracker](https://github.com/SpunkyBot/spunkybot/issues?state=open).

## Contributing

You can help us in different ways:

* Open an [issue](https://github.com/SpunkyBot/spunkybot/issues) with suggestions for improvements
* Fork this repository and submit a pull request:
  * Click the Fork button to create your personal fork
  * Create your feature branch (`git checkout -b new-feature`)
  * Commit your changes (`git commit -am 'Add some feature'`)
  * Push to the branch (`git push origin new-feature`)
  * Create a new pull request
* Improve the [documentation](https://github.com/SpunkyBot/spunkybot-docs) (separate repository)

By contributing code to this project in any form, including sending a pull request via GitHub, a code fragment or patch via mail or public discussion groups, you agree to release your code under the terms of the MIT license that you can find in the [LICENSE](https://github.com/SpunkyBot/spunkybot/blob/develop/LICENSE) file included in this source distribution.

Please see the [CONTRIBUTING](https://github.com/SpunkyBot/spunkybot/blob/develop/.github/CONTRIBUTING.md) guide for information regarding the contribution process.

## License

The code of Spunky Bot is released under the MIT License. See the [LICENSE](https://github.com/SpunkyBot/spunkybot/blob/develop/LICENSE) file for the full license text.

### Third Party Libraries

* RCON: [pyquake3.py](https://github.com/urthub/pyquake3)
  * The library has been modified to fix some error handling issues and fulfill the PEP8 conformance. This file is released under the GNU General Public License.
* GeoIP: [pygeoip.py](https://github.com/urthub/pygeoip)
  * The library has been extended with the list `GeoIP_country_name` to support full country names (e.g. Germany for country_code DE). This file is released under the MIT License.
* GeoLite database: [www.maxmind.com](http://www.maxmind.com)
  * The GeoLite databases created by MaxMind are distributed under the Creative Commons Attribution-ShareAlike 4.0 International License and the MaxMind Open Data License.
* Schedule: [schedule.py](https://github.com/dbader/schedule)
  * This file is released under the MIT License.

Urban Terror® and FrozenSand™ are trademarks, or registered trademarks of Frozensand Games Limited.

## Thank you!

We really appreciate all kinds of feedback and contributions. Thanks for using and supporting Spunky Bot!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/SpunkyBot/spunkybot/compare/1.13.0...develop)

## [1.13.0](https://github.com/SpunkyBot/spunkybot/compare/1.12.2...1.13.0) - 2022-01-16

### Added

* Added command `!lastvote` to display the last called vote
* Added command `!lastadmin` to display the last disconnected admin
* Added option to show number of admins online in rotating messages
* Added Magnum and MAC11 option to command `!gear`
* Added banpoints for racism and camp warning types

### Changed

* Show hint to register for saving player stats
* Show more messages for higher headshot counts
* Show additional hints for command `!exit`
* Show message if no one is listed in the banlist

### Fixed

* Fixed #70: Additional warning types lead to automated temp-ban
* Fixed #75: Warn in logfile if g_logsync is not set correctly
* Fixed #85: Add command `!lastadmin` to display the last disconnected admin
* Fixed #93: Add config item to set which adminlevel is punished for spawn/teamkills
* Fixed #110: Command `!nextmap` supports >52 maps on Windows server
* Fixed #127: Add support for recording serverside demos
* Various minor bug fixes

## [1.12.2](https://github.com/SpunkyBot/spunkybot/compare/1.12.0...1.12.2) - 2020-07-24

### Fixed

* Fixed #61: Ignore mapcycle.txt file in case of using dynamic mapcycle feature
* Fixed #86: Problem with big and small cycle

## [1.12.0](https://github.com/SpunkyBot/spunkybot/compare/1.11.0...1.12.0) - 2020-07-11

### Added

* Added feature for delaying votes after failed vote
* Added messages for cyclemap and kick votes
* Added alias `!kiss` for command `!clear` to clear all player warnings
* Added command `!setgravity` to set the gravity value
* Added alias `!lowgravity on/off` for command `!moon on/off`
* Added alias `!sp` for command `!spec` to move yourself to spectator
* Added command `!mapcycle` to list the map rotation
* Added welcome message for first connection
* Added option to enable/disable resetting of HS and kill spree stats on map rotation
* Added command `!ping <name>` to display the ping of a player

### Changed

* Show First Teamkill message
* Show losing streak and extend `!spree` command
* Show message if no maps have been played when using command `!lastmaps`
* Show message that you cannot ban/kick yourself
* Sync mapcycle file when using command `!rebuild`
* Removed Heartbeat option and online version check from code
* Improved output for successful nextmap vote
* Improved auto-kick warn message for team killing
* Imported latest GeoIP database (January 2020)

### Fixed

* Fixed #60: Fix broken message output for Spawn Killing
* Fixed #64: Handling of bots
* Fixed #65: Nextmap does not show correct value
* Fixed #69: move spawn kill timer into config file
* Various minor bug fixes

## [1.11.0] - 2018-08-06

### Added

* Added command `!banall <pattern>` to ban all players matching pattern
* Added command `!rebuild` to sync up all available maps
* Added option to kill spawnkillers instantly
* Added support for Urban Terror 4.3.4

### Changed

* Command `!maps` is showing the total number of available maps
* Exclude bots from autokick for team kills to avoid unbalanced teams
* Exclude bots from autokick of low score players to avoid unbalanced teams
* Updated schedule library
* Improved some feedback messages

### Fixed

* Fixed #55: Make bots immune from spawnkill autokick
* Fixed #56: Command `!maps` does not show all maps
* Fixed #57: Fix year 2038 problem on 32-bit systems
* Fixed version check, issued by string compare
* Various minor bug fixes

## [1.10.0] - 2018-05-31

### Added

* Added option for automatic expiration of warnings
* Added option to kick players for using bad words
* Added option to display multi-kill messages
* Added option "free" to release lock from forced team
* Added option to define ban duration for command `!ban`
* Added command `!grudge [<name>]` to grudge a player for team killing
* Added command `!gear` to set allowed weapons
* Added command `!forgiveclear <name>` to clear player team kills
* Added command `!forgiveinfo <name>` to display a players team kills
* Added command `!forgivelist` to list players who killed you
* Added command `!forgive [<name>]` to forgive team kills
* Added command `!regulars` to display regular players online
* Added command `!unreg <name>` to remove player from regular group
* Added commands to change the gametype
* Added support for old <ut_name> map names
* Added support for latest Urban Terror 4.3.3 release
* Added support for ioquake3 binary
* Added support for macOS 10.13 and Ubuntu 18.04
* Added reason 'score' to reason dictionary
* Added "Top Objectives" awards for CTF and Bomb mode

### Changed

* Do not show forgive notice for grudged players
* Clear team kills for command `!warnclear`  and `!clear`
* Clear warnings if already expired
* Command `!warninfo` shows now all active warnings
* Command `!swap` moves admin using command if playerB is not given
* Command `!tempban` supports now up 1 sec to 3 days periods
* Show msg "Planted?" when bomb explodes
* Introduced detonation time for bomb mode
* Imported latest GeoIP database (April 2018)
* Unified warn system: warnings + team kills are now combined
* Converted loopback/localhost to 127.0.0.1

### Fixed

* Fixed #50: Improve parsing rotation messages file
* Fixed #49: Ban directly for 15 mins
* Fixed #48: Extend max tempban duration to 3 days
* Fixed #47: Show last maps only for nextmap/map votes
* Fixed #46: Add commands to change the gametype
* Fixed #15: Add command `!grudge [<name>]` to grudge a player for team killing
* Various minor bug fixes

## [1.9.0] - 2017-05-14

### Added

* Added Monster Kill counter

lib/pygeoip.py
#!/usr/bin/env python2.5

'''
Pure Python reader for GeoIP Country Edition databases.
http://python-geoip.googlecode.com/hg/pygeoip.py
'''

__author__ = 'David Wilson <dw@botanicus.net>'

import os

from cStringIO import StringIO

# From GeoIP.h.
SEGMENT_RECORD_LENGTH = 3
STANDARD_RECORD_LENGTH = 3
ORG_RECORD_LENGTH = 4
MAX_RECORD_LENGTH = 4
FULL_RECORD_LENGTH = 50
NUM_DB_TYPES = 20

GEOIP_COUNTRY_EDITION = 1
GEOIP_REGION_EDITION_REV0 = 7
GEOIP_CITY_EDITION_REV0 = 6
GEOIP_ORG_EDITION = 5
GEOIP_ISP_EDITION = 4
GEOIP_CITY_EDITION_REV1 = 2
GEOIP_REGION_EDITION_REV1 = 3
GEOIP_PROXY_EDITION = 8
GEOIP_ASNUM_EDITION = 9
GEOIP_NETSPEED_EDITION = 10
GEOIP_DOMAIN_EDITION = 11
GEOIP_COUNTRY_EDITION_V6 = 12

COUNTRY_BEGIN = 16776960
STATE_BEGIN_REV0 = 16700000
STATE_BEGIN_REV1 = 16000000
STRUCTURE_INFO_MAX_SIZE = 20
DATABASE_INFO_MAX_SIZE = 100

GeoIP_country_code = '''
    AP EU AD AE AF AG AI AL AM AN AO AQ AR AS AT AU AW AZ BA BB BD BE BF BG BH
    BI BJ BM BN BO BR BS BT BV BW BY BZ CA CC CD CF CG CH CI CK CL CM CN CO CR
    CU CV CX CY CZ DE DJ DK DM DO DZ EC EE EG EH ER ES ET FI FJ FK FM FO FR FX
    GA GB GD GE GF GH GI GL GM GN GP GQ GR GS GT GU GW GY HK HM HN HR HT HU ID
    IE IL IN IO IQ IR IS IT JM JO JP KE KG KH KI KM KN KP KR KW KY KZ LA LB LC
    LI LK LR LS LT LU LV LY MA MC MD MG MH MK ML MM MN MO MP MQ MR MS MT MU MV
    MW MX MY MZ NA NC NE NF NG NI NL NO NP NR NU NZ OM PA PE PF PG PH PK PL PM
    PN PR PS PT PW PY QA RE RO RU RW SA SB SC SD SE SG SH SI SJ SK SL SM SN SO
    SR ST SV SY SZ TC TD TF TG TH TJ TK TM TN TO TL TR TT TV TW TZ UA UG UM US
    UY UZ VA VC VE VG VI VN VU WF WS YE YT RS ZA ZM ME ZW A1 A2 O1 AX GG IM JE
    BL MF
'''.split()

GeoIP_country_name = (
    'Asia/Pacific Region', 'Europe', 'Andorra', 'United Arab Emirates',
    'Afghanistan', 'Antigua and Barbuda', 'Anguilla', 'Albania', 'Armenia',
    'Netherlands Antilles', 'Angola', 'Antarctica', 'Argentina',
    'American Samoa', 'Austria', 'Australia', 'Aruba', 'Azerbaijan',
    'Bosnia and Herzegovina', 'Barbados', 'Bangladesh', 'Belgium',
    'Burkina Faso', 'Bulgaria', 'Bahrain', 'Burundi', 'Benin', 'Bermuda',
    'Brunei Darussalam', 'Bolivia', 'Brazil', 'Bahamas', 'Bhutan',
    'Bouvet Island', 'Botswana', 'Belarus', 'Belize', 'Canada',
    'Cocos (Keeling) Islands', 'Congo, The Democratic Republic of the',
    'Central African Republic', 'Congo', 'Switzerland', 'Cote D\'Ivoire',
    'Cook Islands', 'Chile', 'Cameroon', 'China', 'Colombia', 'Costa Rica',
    'Cuba', 'Cape Verde', 'Christmas Island', 'Cyprus', 'Czech Republic',
    'Germany', 'Djibouti', 'Denmark', 'Dominica', 'Dominican Republic',
    'Algeria', 'Ecuador', 'Estonia', 'Egypt', 'Western Sahara', 'Eritrea',
    'Spain', 'Ethiopia', 'Finland', 'Fiji', 'Falkland Islands (Malvinas)',
    'Micronesia, Federated States of', 'Faroe Islands', 'France',
    'France, Metropolitan', 'Gabon', 'United Kingdom', 'Grenada', 'Georgia',
    'French Guiana', 'Ghana', 'Gibraltar', 'Greenland', 'Gambia', 'Guinea',
    'Guadeloupe', 'Equatorial Guinea', 'Greece',
    'South Georgia and the South Sandwich Islands', 'Guatemala', 'Guam',
    'Guinea-Bissau', 'Guyana', 'Hong Kong',
    'Heard Island and McDonald Islands', 'Honduras', 'Croatia', 'Haiti',
    'Hungary', 'Indonesia', 'Ireland', 'Israel', 'India',
    'British Indian Ocean Territory', 'Iraq', 'Iran, Islamic Republic of',
    'Iceland', 'Italy', 'Jamaica', 'Jordan', 'Japan', 'Kenya', 'Kyrgyzstan',
    'Cambodia', 'Kiribati', 'Comoros', 'Saint Kitts and Nevis',
    'Korea, Democratic People\'s Republic of', 'Korea, Republic of', 'Kuwait',
    'Cayman Islands', 'Kazakhstan', 'Lao People\'s Democratic Republic',
    'Lebanon', 'Saint Lucia', 'Liechtenstein', 'Sri Lanka', 'Liberia',
    'Lesotho', 'Lithuania', 'Luxembourg', 'Latvia', 'Libya', 'Morocco',
    'Monaco', 'Moldova, Republic of', 'Madagascar', 'Marshall Islands',
    'Macedonia', 'Mali', 'Myanmar', 'Mongolia', 'Macau',
    'Northern Mariana Islands', 'Martinique', 'Mauritania', 'Montserrat',
    'Malta', 'Mauritius', 'Maldives', 'Malawi', 'Mexico', 'Malaysia',
    'Mozambique', 'Namibia', 'New Caledonia', 'Niger', 'Norfolk Island',
    'Nigeria', 'Nicaragua', 'Netherlands', 'Norway', 'Nepal', 'Nauru', 'Niue',
    'New Zealand', 'Oman', 'Panama', 'Peru', 'French Polynesia',
    'Papua New Guinea', 'Philippines', 'Pakistan', 'Poland',
    'Saint Pierre and Miquelon', 'Pitcairn Islands', 'Puerto Rico',
    'Palestinian Territory', 'Portugal', 'Palau', 'Paraguay', 'Qatar',
    'Reunion', 'Romania', 'Russian Federation', 'Rwanda', 'Saudi Arabia',
    'Solomon Islands', 'Seychelles', 'Sudan', 'Sweden', 'Singapore',
    'Saint Helena', 'Slovenia', 'Svalbard and Jan Mayen', 'Slovakia',
    'Sierra Leone', 'San Marino', 'Senegal', 'Somalia', 'Suriname',
    'Sao Tome and Principe', 'El Salvador', 'Syrian Arab Republic',
    'Swaziland', 'Turks and Caicos Islands', 'Chad',
    'French Southern Territories', 'Togo', 'Thailand', 'Tajikistan', 'Tokelau',
    'Turkmenistan', 'Tunisia', 'Tonga', 'Timor-Leste', 'Turkey',
    'Trinidad and Tobago', 'Tuvalu', 'Taiwan', 'Tanzania, United Republic of',
    'Ukraine', 'Uganda', 'United States Minor Outlying Islands',
    'United States', 'Uruguay', 'Uzbekistan', 'Holy See (Vatican City State)',
    'Saint Vincent and the Grenadines', 'Venezuela', 'Virgin Islands, British',
    'Virgin Islands, U.S.', 'Vietnam', 'Vanuatu', 'Wallis and Futuna', 'Samoa',
    'Yemen', 'Mayotte', 'Serbia', 'South Africa', 'Zambia', 'Montenegro',
    'Zimbabwe', 'Anonymous Proxy', 'Satellite Provider', 'Other',
    'Aland Islands', 'Guernsey', 'Isle of Man', 'Jersey', 'Saint Barthelemy',
    'Saint Martin', 'Bonaire, Sint Eustatius and Saba', 'South Sudan'
)

GeoIP_country_continent = '''
    AS EU EU AS AS SA SA EU AS SA AF AN SA OC EU OC SA AS EU SA AS EU AF EU AS
    AF AF SA AS SA SA SA AS AF AF EU SA NA AS AF AF AF EU AF OC SA AF AS SA SA
    SA AF AS AS EU EU AF EU SA SA AF SA EU AF AF AF EU AF EU OC SA OC EU EU EU
    AF EU SA AS SA AF EU SA AF AF SA AF EU SA SA OC AF SA AS AF SA EU SA EU AS
    EU AS AS AS AS AS EU EU SA AS AS AF AS AS OC AF SA AS AS AS SA AS AS AS SA
    EU AS AF AF EU EU EU AF AF EU EU AF OC EU AF AS AS AS OC SA AF SA EU AF AS
    AF NA AS AF AF OC AF OC AF SA EU EU AS OC OC OC AS SA SA OC OC AS AS EU SA
    OC SA AS EU OC SA AS AF EU AS AF AS OC AF AF EU AS AF EU EU EU AF EU AF AF
    SA AF SA AS AF SA AF AF AF AS AS OC AS AF OC AS AS SA OC AS AF EU AF OC NA
    SA AS EU SA SA SA SA AS OC OC OC AS AF EU AF AF EU AF -- -- -- EU EU EU EU
    SA SA
'''.split()


# Helper functions

def addr_to_num(ip):
    '''
    Convert an IPv4 address from a string to its integer representation.

    @param[in]  ip      IPv4 address as a string.
    @returns            Address as an integer.
    '''

    try:
        w, x, y, z = map(int, ip.split('.'))
        if w > 255 or x > 255 or y > 255 or z > 255:
            raise ValueError()
    except ValueError:
        raise ValueError('%r is not an IPv4 address.' % (ip,))

    return (w << 24) | (x << 16) | (y << 8) | z


def num_to_addr(num):
    '''
    Convert an IPv4 address from its integer representation to a string.

    @param[in]  num     Address as an integer.
    @returns            IPv4 address as a string.
    '''

    return '%d.%d.%d.%d' % ((num >> 24) & 0xff,
                            (num >> 16) & 0xff,
                            (num >> 8) & 0xff,
                            (num & 0xff))


def latin1_to_utf8(string):
    return string.decode('latin-1').encode('utf-8')


def safe_lookup(lst, idx):
    if idx is None:
        return None
    return lst[idx]


class ReadBuffer(object):
    '''
    Utility to read data more easily.
    '''
    buffer = None

    def __init__(self, source, size, seek_offset=None, seek_whence=os.SEEK_SET):
        fp = StringIO(source)
        if seek_offset is not None:
            fp.seek(seek_offset, seek_whence)
        self.buffer = fp.read(size)

    def read_string(self):
        '''
        Read a null-terminated string.

        @returns            Result as a string.
        '''
        result, self.buffer = self.buffer.split('\0', 1)
        return result

    def read_int(self, size):
        '''
        Read a multibyte integer.

        @param[in]  size    Number of bytes to read as an integer.
        @returns            Result as an integer.
        '''
        result = sum(ord(self.buffer[i]) << (8 * i) for i in range(size))
        self.buffer = self.buffer[size:]
        return result


class AddressInfo(object):
    '''
    Representation of a database lookup result.
    '''
    __slots__ = ['ip', 'ipnum', 'prefix', 'country', 'country_name', 'continent']

    def __init__(self, ip=None, ipnum=None, prefix=None, country_id=None):
        self.ip = ip
        self.ipnum = ipnum
        self.prefix = prefix
        self.country = safe_lookup(GeoIP_country_code, country_id)
        self.country_name = safe_lookup(GeoIP_country_name, country_id)
        self.continent = safe_lookup(GeoIP_country_continent, country_id)

    network = property(lambda self:
                       num_to_addr(self.ipnum & ~((32 - self.prefix) ** 2 - 1)))

    def __str__(self):
        return '[%s of network %s/%d in country %s]' % \
               (self.ip, self.network, self.prefix, self.country)


class BigAddressInfo(AddressInfo):
    '''
    Representation of a database lookup result with more info in it.
    '''
    # __slots__ is inherited and appended to.
    __slots__ = ['city', 'region', 'postal_code', 'metro_code', 'area_code', 'longitude', 'latitude']

    def __init__(self, ip=None, ipnum=None, prefix=None, country_id=None,
                 city=None, region=None, postal_code=None, metro_code=None,
                 area_code=None, longitude=None, latitude=None):
        AddressInfo.__init__(self, ip, ipnum, prefix, country_id)
        self.city = city or None
        self.region = region or None
        self.postal_code = postal_code or None
        self.metro_code = metro_code
        self.area_code = area_code
        self.longitude = longitude
        self.latitude = latitude

    def __str__(self):
        return '[%s of network %s/%d in city %s, %s]' % \
               (self.ip, self.network, self.prefix, self.city, self.country)


class Database(object):
    '''
    GeoIP database reader implementation. Currently only supports country
    edition.
    '''

    def __init__(self, filename):
        '''
        Initialize a new GeoIP reader instance.

        @param[in]  filename    Path to GeoIP.dat as a string.
        '''
        self.filename = filename
        self.segments = None
        self.db_type = None
        self.record_length = None
        self.cache = open(filename, 'rb').read()
        self._setup_segments()

        if self.db_type not in (GEOIP_COUNTRY_EDITION,
                                GEOIP_CITY_EDITION_REV0,
                                GEOIP_CITY_EDITION_REV1):
            raise NotImplementedError('Database edition is not supported yet; '
                                      'Please use a Country or City database.')

    def _setup_segments(self):
        # default to GeoIP Country Edition
        self.db_type = GEOIP_COUNTRY_EDITION
        self.record_length = STANDARD_RECORD_LENGTH

        fp = StringIO(self.cache)
        fp.seek(-3, os.SEEK_END)

        for i in range(STRUCTURE_INFO_MAX_SIZE):
            delim = fp.read(3)

            if delim != '\xFF\xFF\xFF':
                fp.seek(-4, os.SEEK_CUR)
                continue

            self.db_type = ord(fp.read(1))

            # Region Edition, pre June 2003.
            if self.db_type == GEOIP_REGION_EDITION_REV0:
                self.segments = [STATE_BEGIN_REV0]

            # Region Edition, post June 2003.
            elif self.db_type == GEOIP_REGION_EDITION_REV1:
                self.segments = [STATE_BEGIN_REV1]

            # City/Org Editions have two segments, read offset of second segment
            elif self.db_type in (GEOIP_CITY_EDITION_REV0,
                                  GEOIP_CITY_EDITION_REV1,
                                  GEOIP_ORG_EDITION, GEOIP_ISP_EDITION,
                                  GEOIP_ASNUM_EDITION):
                self.segments = [0]

                for idx, ch in enumerate(fp.read(SEGMENT_RECORD_LENGTH)):
                    self.segments[0] += ord(ch) << (idx * 8)

                if self.db_type in (GEOIP_ORG_EDITION, GEOIP_ISP_EDITION):
                    self.record_length = ORG_RECORD_LENGTH

            break

        if self.db_type in (GEOIP_COUNTRY_EDITION, GEOIP_PROXY_EDITION,
                            GEOIP_NETSPEED_EDITION, GEOIP_COUNTRY_EDITION_V6):
            self.segments = [COUNTRY_BEGIN]

    def info(self):
        '''
        Return a string describing the loaded database version.

        @returns    English text string, or None if database is ancient.
        '''
        fp = StringIO(self.cache)
        fp.seek(-3, os.SEEK_END)

        hasStructureInfo = False

        # first get past the database structure information
        for i in range(STRUCTURE_INFO_MAX_SIZE):
            if fp.read(3) == '\xFF\xFF\xFF':
                hasStructureInfo = True
                break

            fp.seek(-4, os.SEEK_CUR)

        if hasStructureInfo:
            fp.seek(-6, os.SEEK_CUR)
        else:
            # no structure info, must be pre Sep 2002 database, go back to end.
            fp.seek(-3, os.SEEK_END)

        for i in range(DATABASE_INFO_MAX_SIZE):
            if fp.read(3) == '\0\0\0':
                return fp.read(i)

            fp.seek(-4, os.SEEK_CUR)

    def _decode(self, buf, branch):
        '''
        @param[in]  buf         Record buffer.
        @param[in]  branch      1 for left, 2 for right.
        @returns                X.
        '''
        offset = 3 * branch
        if self.record_length == 3:
            return buf[offset] | (buf[offset + 1] << 8) | (buf[offset + 2] << 16)

        # General case.
        end = branch * self.record_length
        x = 0

        for j in range(self.record_length):
            x = (x << 8) | buf[end - j]

        return x

    def _seek_record(self, ipnum):
        fp = StringIO(self.cache)
        offset = 0

        for depth in range(31, -1, -1):
            fp.seek(self.record_length * 2 * offset)
            buf = map(ord, fp.read(self.record_length * 2))

            x = self._decode(buf, int(bool(ipnum & (1 << depth))))
            if x >= self.segments[0]:
                return 32 - depth, x

            offset = x

        assert False, \
            "Error Traversing Database for ipnum = %lu: "\
            "Perhaps database is corrupt?" % ipnum

    def _lookup_country(self, ip):
        "Lookup a country db entry."

        ipnum = addr_to_num(ip)
        prefix, num = self._seek_record(ipnum)

        num -= COUNTRY_BEGIN
        if num:
            country_id = num - 1
        else:
            country_id = None

        return AddressInfo(country_id=country_id, ip=ip, ipnum=ipnum, prefix=prefix)

    def _lookup_city(self, ip):
        "Look up a city db entry."

        ipnum = addr_to_num(ip)
        prefix, num = self._seek_record(ipnum)
        record, next_record_ptr = self._extract_record(num, None)
        return BigAddressInfo(ip=ip, ipnum=ipnum, prefix=prefix, **record)

    def _extract_record(self, seek_record, next_record_ptr):
        if seek_record == self.segments[0]:
            return {'country_id': None}, next_record_ptr

        seek_offset = seek_record + (2 * self.record_length - 1) * self.segments[0]
        record_buf = ReadBuffer(self.cache, FULL_RECORD_LENGTH, seek_offset)
        record = {}

        # get country
        record['country_id'] = record_buf.read_int(1) - 1

        # get region
        record['region'] = record_buf.read_string()

        # get city
        record['city'] = latin1_to_utf8(record_buf.read_string())

        # get postal code
        record['postal_code'] = record_buf.read_string()

        # get latitude
        record['latitude'] = record_buf.read_int(3) / 10000.0 - 180

        # get longitude
        record['longitude'] = record_buf.read_int(3) / 10000.0 - 180

        # get area code and metro code for post April 2002 databases and for
        # US locations
        if (self.db_type == GEOIP_CITY_EDITION_REV1) and (GeoIP_country_code[record['country_id']] == 'US'):
            metro_area_combo = record_buf.read_int(3)
            record['metro_code'] = metro_area_combo / 1000
            record['area_code'] = metro_area_combo % 1000

        # Used for GeoIP_next_record (which this code doesn't have.)
        if next_record_ptr is not None:
            next_record_ptr = seek_record - len(record_buf)

        return record, next_record_ptr

    def lookup(self, ip):
        '''
        Lookup an IP address returning an AddressInfo (or BigAddressInfo)
        instance describing its location.

        @param[in]  ip      IPv4 address as a string.
        @returns            AddressInfo (or BigAddressInfo) instance.
        '''

        if self.db_type in (GEOIP_COUNTRY_EDITION, GEOIP_PROXY_EDITION, GEOIP_NETSPEED_EDITION):
            return self._lookup_country(ip)
        elif self.db_type in (GEOIP_CITY_EDITION_REV0, GEOIP_CITY_EDITION_REV1):
            return self._lookup_city(ip)
